tools/backups2datalad/zarr.py

```diff
@@ -0,0 +1,405 @@
+from __future__ import annotations
+
+from collections import deque
+from dataclasses import dataclass, field
+from datetime import datetime
+import os
+from pathlib import Path
+import sys
+from typing import AsyncIterator, Optional, Set, Tuple
+from urllib.parse import urlparse, urlunparse
+
+import boto3
+from botocore import UNSIGNED
+from botocore.client import Config as BotoConfig
+from dandi.dandiapi import RemoteZarrAsset, RemoteZarrEntry, ZarrEntryStat, ZarrListing
+from datalad.api import Dataset
+import dateutil.parser
+import httpx
+import trio
+
+from .annex import AsyncAnnex
+from .consts import ZARRS_REMOTE_PREFIX, ZARRS_REMOTE_UUID
+from .util import (
+    Config,
+    MiniFuture,
+    Remote,
+    arequest,
+    create_github_sibling,
+    custom_commit_date,
+    init_dataset,
+    key2hash,
+    log,
+    maxdatetime,
+    quantify,
+)
+
+if sys.version_info[:2] >= (3, 10):
+    from contextlib import aclosing
+else:
+    from async_generator import aclosing
+
+
+CHECKSUM_FILE = ".zarr-checksum"
+
+
+@dataclass
+class ZarrReport:
+    added: int = 0
+    updated: int = 0
+    deleted: int = 0
+    checksum: bool = False
+
+    def __bool__(self) -> bool:
+        return bool(self.added or self.updated or self.deleted or self.checksum)
+
+    def get_summary(self) -> str:
+        msgparts = []
+        if self.added:
+            msgparts.append(f"{quantify(self.added, 'file')} added")
+        if self.updated:
+            msgparts.append(f"{quantify(self.updated, 'file')} updated")
+        if self.deleted:
+            msgparts.append(f"{quantify(self.deleted, 'file')} deleted")
+        if self.checksum:
+            msgparts.append("checksum updated")
+        if not msgparts:
+            msgparts.append("No changes")
+        return ", ".join(msgparts)
+
+
+@dataclass
+class ZarrSyncer:
+    api_url: str
+    zarr_id: str
+    repo: Path
+    annex: AsyncAnnex
+    s3client: httpx.AsyncClient
+    backup_remote: Optional[str]
+    checksum: str
+    last_timestamp: Optional[datetime] = None
+    extant_paths: Set[str] = field(default_factory=set)
+    report: ZarrReport = field(default_factory=ZarrReport)
+
+    async def run(self) -> None:
+        async with aclosing(self.aiter_file_entries()) as ait:  # type: ignore[type-var]
+            async for entry in ait:
+                log.info("Zarr %s: %s: Syncing", self.zarr_id, entry)
+                self.extant_paths.add(str(entry))
+                bucket_url, st = await self.stat(entry)
+                md5_digest = entry.get_digest().value
+                self.last_timestamp = maxdatetime(self.last_timestamp, st.modified)
+                dest = self.repo / str(entry)
+                if dest.is_dir():
```

Comment:
me and future me would like to have a comment here above on why are we to remove something locally, e.g. here (I guess) because listing S3  would not return us names of directories, if we encounter directory locally it means that a directory (prefix used by multiple keys before on S3) turned into a file (on S3).

Reply:
What exactly are you asking for a comment to explain?

Reply:
smth like this
```suggestion
                # entries (from S3) can contain be files, not directories.
                # We might need first to remove conflicts locally first to place those files.
                if dest.is_dir():
```
diff is odd -- might have screwed up indentation

Reply:
I added comments in d6e21c0.